draft-barnes-mls-appsync.md

```diff
@@ -60,171 +61,225 @@ communicate. Applications based on MLS can integrate their state into this
+amounts of data for a small change.
 
-1. A GroupContext extension `application_states` that confirms agreement on
-   application state from potentially multiple sources.
-2. A new proposal type AppSync that allows MLS group members to propose changes
-   to the agreed application state.
+Here, we define a new MLS proposal type GroupContextExtensionsDiff. This proposal type allows modification on a per extension basis. It also allows GroupContext extensions to define simple diff formats (for example, to add, remove, and update elements in a list or items in a map).
 
 # Conventions and Definitions
 
 {::boilerplate bcp14-tagged}
 
-# Application State Synchronization
+This document uses many terms from {{!RFC9420}}. Some of these have
+names which are similar or may be confusing: GroupContext, GroupContext
+extension, ExtensionType, and GroupContextExtensions proposal.
 
-This document defines a new AppSync proposal. AppSync is a Safe Extension as
-defined in {{Section 2 of !I-D.ietf-mls-extensions}}, of type
-`extension_external_proposal`.
+# GroupContext extensions modification
 
-The `application_states` extension allows the application to inject state
-objects into the MLS key schedule. Changes to this state can be made out of
-band, or using the AppSync proposal. Using the AppSync proposal ensures that
-members of the MLS group have received the relevant state changes before they
-are reflected in the group's `application_states`.
-
-> **NOTE:** This design exposes the high-level structure of the application state
-> to MLS.  An alternative design would be to have the application state be opaque
-> to MLS.  There is a trade-off between generality and the complexity of the API
-> between the MLS implementation and the application.  An opaque design would give
-> the application more freedom, but require the MLS stack to call out to the
-> application to get the updated state as part of Commit processing.  This design
-> allows the updates to happen within the MLS stack, so that no callback is
-> needed, at the cost of forcing the application state to fit a certain structure.
-> It also potentially can result in smaller state updates in large groups.
-
-The state for Each `applicationId` in the `application_states` needs to conform
-to one of four basic types: an ordered array, an unordered array, a map, or an
-irreducible blob. This allows the AppSync proposal to efficiently modify a large
-application state object.
-
-The content of the `application_states` extension and the `AppSync` proposal are
-structured as follows:
+This document defines a new GroupContextExtensionsDiff proposal. It
+is not a Safe Extension as defined in {{Section 2 of
+!I-D.ietf-mls-extensions}}, because it can modify any GroupContext
+extension, including those defined in {{!RFC9420}}. It is intended as a
+complete replacement for the GroupContextExtensions Proposal.
+The GroupContextExtensionsDiff proposal does not require an UpdatePath,
```

Comment:
GCE requires an UpdatePath, and this is strictly stronger than GCE, so it seems like it has to require an UpdatePath.

Reply:
That was listed as one of the reasons why GCE was broken AFAIK.